### PR TITLE
CNF-12621: Add Assisted Installer reference templates

### PR DESCRIPTION
**Pull request by sakhoury**
This PR contains the following additions:
- Assisted Installer reference templates
- Functionality to create the reference ConfigMaps on initialization of the SiteConfig operator.

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -136,3 +152,31 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+func initConfigMapTemplates(ctx context.Context, c client.Client, log logr.Logger) error {
```
Comment (eranco74) [inline cmd/main.go:156]: Consider renaming to initAssistedInstallerConfigMapTemplates

Comment (sakhoury) [reply cmd/main.go:156]: Subsequent to this PR, I will be creating another one for IBI templates. This function will be used to apply those templates as well. Hence, the generic name.

File: internal/templates/assisted-installer/template.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    disableCertificateVerification: true
+    credentialsName: "{{ .SpecialVars.CurrentNode.BmcCredentialsName.Name }}"
+  bootMACAddress: "{{ .SpecialVars.CurrentNode.BootMACAddress }}"
+  automatedCleaningMode: "{{ .SpecialVars.CurrentNode.AutomatedCleaningMode }}"
+  online: true
+{{ if (anyFieldDefined .SpecialVars.CurrentNode.RootDeviceHints) }}
+  rootDeviceHints:
+{{ .SpecialVars.CurrentNode.RootDeviceHints | toYaml | indent 4 }}
+{{ end }}`
+
```
Comment (imiller0) [inline internal/templates/assisted-installer/template.go:199]: Original siteConfig had the userData field. Is this no longer supported?

Comment (imiller0) [inline internal/templates/assisted-installer/template.go:199]: There is an optional CR for HostFirmwareSettings. The feature for it is slated to GA in 4.17. We should make sure it remains supported by default. ie if the user has the field in the SiteConfig we should generate the CR.

File: internal/templates/assisted-installer/template.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+const AgentClusterInstall = `apiVersion: extensions.hive.openshift.io/v1beta1
+kind: AgentClusterInstall
+metadata:
+  name: "{{ .Site.ClusterName }}"
+  namespace: "{{ .Site.ClusterName }}"
+  annotations:
+{{ if .SpecialVars.InstallConfigOverrides }}
+    agent-install.openshift.io/install-config-overrides: '{{ .SpecialVars.InstallConfigOverrides }}'
+{{ end }}
+    metaclusterinstall.openshift.io/sync-wave: "1"
```
[the remark this answers is not in the document]
Comment (sakhoury) [reply internal/templates/assisted-installer/template.go:27]: They are used for creation. Certain manifests need to be created before others.

File: internal/templates/assisted-installer/template.go (hunk 2 above)
Comment (sakhoury) [reply internal/templates/assisted-installer/template.go:199]: It was removed. Discussion around the `userData` field: https://github.com/sakhoury/siteconfig/pull/1#discussion_r1561578977.

Comment (sakhoury) [reply internal/templates/assisted-installer/template.go:199]: I can include the HostFirmwareSettings CR as part of the default set for Assisted Installer, but what would the spec field reference?
```yaml
apiVersion: metal3.io/v1alpha1
kind: HostFirmwareSettings
metadata:
  annotations:
    metaclusterinstall.openshift.io/sync-wave: "1"
  name: "{{ .SpecialVars.CurrentNode.HostName }}"
  namespace: "{{ .Site.ClusterName }}"
spec: "{{ ??? }}"
```

Comment (sabbir-47) [reply internal/templates/assisted-installer/template.go:199]: @sakhoury For HP servers, we have below recommendations, i don't know that can be helpful for you. Please ignore anything above spec.
```yaml
apiVersion: metal3.io/v1alpha1
kind: HostFirmwareSettings
metadata:
    annotations:
        argocd.argoproj.io/sync-wave: "1"
        ran.openshift.io/ztp-gitops-generated: '{}'
    name: node1
    namespace: cluster1
spec:
    settings:
        BootMode: Uefi
        BootOrderPolicy: RetryIndefinitely
        ProcHyperthreading: Enabled
        ProcVirtualization: Enabled
        UefiOptimizedBoot: Enabled
        UefiSerialDebugLevel: Disabled
```

Comment (sakhoury) [reply internal/templates/assisted-installer/template.go:199]: Thanks @sabbir-47. Is this specific to only HP servers? If so, the user can always add a HostFirmwareSettings template specific to their servers (cc @imiller0).

**Review by eranco74 — APPROVED** [review]

**Pull request merged**